fastlane/Fastfile

```diff
@@ -584,28 +584,59 @@ end
 
 desc "Ensure the correct formatting for the fastlane tools"
 private_lane :ensure_tool_name_formatting do
+  test_tool_name_formatting_ensurance_lane
   UI.message("🕗  Verifying tool name formatting...")
   require 'fastlane/tools'
   errors = []
   Dir.chdir("..") do
     Dir["**/*.md"].each do |path|
       content = File.read(path)
-      Fastlane::TOOLS.each do |tool|
-        errors << "Use _#{tool}_ instead of `#{tool}` to mention a tool in the docs in '#{path}'" if content.include?("`#{tool}`")
-        errors << "Use _#{tool}_ instead of `_#{tool}_` to mention a tool in the docs in '#{path}'" if content.include?("`_#{tool}_`")
-        errors << "Use [_#{tool}_] instead of [#{tool}] to mention a tool in the docs in '#{path}'" if content.include?("[#{tool}]")
-        errors << "Use <em>#{tool}<em> instead of <code>#{tool}</code> to mention a tool in the docs in '#{path}'" if content.include?("<code>#{tool}</code>")
-        if content.include?("_#{tool.to_s.capitalize}_") || content.include?("`#{tool.to_s.capitalize}`")
-          errors << "fastlane tools have to be formatted in lower case: #{tool} in '#{path}'"
-        end
-      end
+      errors += find_tool_name_formatting_errors(content: content, path: path)
     end
   end
   errors.each { |a| UI.error(a) }
   UI.user_error!("Invalid formatting of one of the fastlane tools") unless errors.empty?
   UI.success("✅  fastlane tools formatting is correct")
 end
 
+private_lane :test_tool_name_formatting_ensurance_lane do
+  content = File.read("spec/fixtures/fastfiles/tool_name_formatting.txt")
+  errors = find_tool_name_formatting_errors(content: content, path: "CONTRIBUTING.md")
+  errors.select! { |e| !e.end_with?("❌") }
```

Comment:
The issue with this is that you're only checking if lines not ending with ❌  are _not_ detected at errors (actually, talk about double-negation… had it wrong the first time I read that code and thought it was doing the opposite at first…), but you're not checking if _all_ the lines ending with ✅  are detected as errors.

If you rewrite this as a spec (as suggested in my top-level comment from a couple of minutes ago), you should `expect(errors).to equal(the_exact_expected_list)`, as opposed to do it similar to what you did it, only expecting that all `errors` end with ✅  but without checking that you didn't miss one of them in the original fixture.

Reply:
Your suggestion for helpers (similar to plugins) makes sense! It's literally the only file we have in the helper folder, nice 🤓 so this can definitely be moved to rspecs and proper unit tests be written, but can you also help me understand what you meant in this message you sent above? I think the tests are covering all lines, whether they're successful or not 🤔 We do check if the ones with ✅ succeed (and they do), and we also check if the ones with ❌ fail (and they do). What's missing? (just trying to understand here, if we were to keep it, although we won't 😇 )

Reply:
Well another way of seeing this is that there are 4 categories of lines:
1. Lines that should be considered errors, and are properly flagged by `find_tool_name_formatting_errors` as such
2. Lines that should be valid, and are properly _not_ flagged by `find_tool_name_formatting_errors` as errors
3. Lines that should be considered errors, but are not properly detected and returned by `find_tool_name_formatting_errors`
4. Lines that should be valid, but are incorrectly flagged as errors by `find_tool_name_formatting_errors`

What we want is to ensure not only that cases 1 and 2 happen (i.e. our implementation work when it's supposed to) but also that cases 3 and 4 don't happen (i.e. that our implementation doesn't create neither false positives or false negatives)

In your code above, you only look at if the return value of `find_tool_name_formatting_errors` (aka `errors`), so you're only considering cases 1 and 4 (i.e. things that have been reported as `errors`). And especially then you're raising a `UI.error` if case 4 (the thing that was raised as errors shouldn't have been) happens.

But you're not checking case 3 at all, i.e. lines that should be considered as errors and should have been returned in `errors` when you called `find_tool_name_formatting_errors`, but weren't reported (in case our implementation of `find_tool_name_formatting_errors` would be incorrect — which is what we want to test here after all)

Reply:
Good point! 🎯